### ISSUE TITLE
[Hardening] Create a mechanism to periodically audit Firecracker's crate dependencies

### ISSUE THREAD

**Issue by raduweiss** [firecracker-microvm/firecracker#1421]
As a security-risk reduction measure, introduce a mechanism to audit Firecracker's crate dependency tree.

While `cargo-audit` notifies us of security issues when they are published, we can avoid exposure by being careful with how much we import.

(note: This doesn't mean we need to re-implement existing functionality)

**Comment by petreeftime** [comment]
I disagree with the statement of this goal, but not the ideal behind it. I don't think it's a generally a good idea to reinvent the wheel. There are a lot of upstream crates that are well maintained and well utilized and the code quality for those will likely be better than their replacement. In case they introduce too many other dependencies, maybe it's a better idea to work with the maintainers of those crates to introduce feature flags where appropriate and have the whole community benefit from the work instead of creating code no one else will end up using.

**Comment by andreeaflorescu** [comment]
@petreeftime It's not only about feature flags here. It's not only about additional features we don't need, it's also about support for operating systems and platforms that we are not interested in.

I also don't believe we want to reinvent the wheel. That's why this issue is formulated as: "only depend on crates maintained by the Firecracker maintainers, by the rust-lang team, and by rust-vmm". It doesn't say **no dependencies**. I really like the CrosVM model where they strive to keep the dependency list to bare minimal. When you don't have 100 dependencies, it is also much easier to audit them.

You will probably not see the Firecracker team re-implement serde, because it does not make sense. But we just need to be more careful when adding dependencies and also we should see if there are easy ones that we can get rid of.

**Comment by dianpopa** [comment]
This was implemented by periodically running cargo-audit against Firecracker.

**Comment by alindima** [comment]
Hmm @dianpopa, AFAIK this issue was tracking a mechanism for auditing the dependency tree in terms of removing dependencies that are not absolutely necessary in order to minimise the amount of external code we use.

Running cargo-audit against Firecracker is only warning us of any known security vulnerabilities.

@raduweiss is that right? should we keep this issue open?

**Comment by raduweiss** [comment]
Just got around to reading this now. @alindima that's right. I'm reopening this since the the intention was also to look at our dependency tree and see if there are sections that we can do without.
